### ISSUE TITLE
Remove canonifyURLs support

### ISSUE THREAD

**Issue by bep** [gohugoio/hugo#4733]
See https://github.com/gohugoio/hugoDocs/issues/469

This is also related to #4574 -- and we should make sure to do this and that in the same release.

We currently have a fair amount of (some of it slowish) code to support `canonifyURLs`, and there are several issues with it. Issues that are very hard to fix in a simple way. I don't use `canonifyURLs`, and I'm not willing to spend time fixing/maintaining it. 

So until someone steps up to the plate with some magic, we are removing it.

/cc @spf13

**Comment by kaushalmodi** [comment]
Currently sites that have `baseURLs` with subdirs **have to** use `canonifyURLs = true` as a crutch to make links like `![](/images/gopher.png)` work for such sites.

See [this test page][1] (*best viewed in widescreen so that the nested frame can be seen side-by-side*).

### The said example image link works with `canonifyURLs = true`

![image](https://user-images.githubusercontent.com/3578197/40235323-ece54fe0-5a77-11e8-9beb-ea2825eb6386.png)

### The said example image link fails with `canonifyURLs = false` (default)

![image](https://user-images.githubusercontent.com/3578197/40235335-f743d830-5a77-11e8-9a35-878b2b0be9cf.png)

[1]: https://hugo-sandbox.netlify.com/subdir-canonify-true/

**Comment by bep** [comment]
@kaushalmodi #4457 should solve the Markdown part. 

But why do you **have** to use canonifyURLs = true to get your example working? Note the boldness in my has. There is an obvious workaround when you say that _an URL in Markdown with a leading slash is relative to the host_. You have a similar challenge with languages.

So, `/subpath/images/gopher.png` isn't very portable, but it works.

 #4457 solves the Markdown part for both sub-paths and languages in a much cleaner and effective way.

So what you lose by this is the canonification of the URLs inside the templates. And that, my friend, is work for `absURL`.

**Comment by kaushalmodi** [comment]
> So, `/subpath/images/gopher.png` isn't very portable

Exactly! I am aware of that workaround.. but portability is a big reason I am not going that route. Tomorrow if a site moves from subdir-domain to regular domain, the needed refactoring, while possible, is still an ugly job.

> And you can create yourelf a shortcode if you really want.

I prefer to use regular Markdown links as much as possible.

> #4457 solves the Markdown part for both sub-paths and languages in a much cleaner and effective way.

I'll subscribe to that issue (*You obviously meant https://github.com/gohugoio/hugo/issues/4574 :)*). I'll definitely try it out once you have a PR for that.

> So what you lose by this is the canonification of the URLs inside the templates. And that, my friend, is work for `absURL`.

Yes, I am fine with that. The real-world usage is mainly linking to figures in content Markdown files, which look like the example I posted above.

**Comment by Jos512** [comment]
I use `canonifyURLs` on my websites to translate relative URLs to absolute ones. But I understand that this feature will be possible with #4574 too? In that case I don't care, since the same functionality that I currently use will be present in the future.

**Comment by bolnel** [comment]
I included canonifyURLs because my images weren't showing when I set up the GitHub page. However, after some tweaking, the images did show when I removed it, so issue must've been somewhere else (maybe because I initially didn't put the trailing slash in baseURL). But there was another issue when I removed it - my menu URLs ended up in the format: customDomain/REPOSITORY_NAME/slug, so they were broken.

I added canonifyURLs again, and the repo name disappeared from the URLs. Not sure what the issue is - I checked a lot of forum threads today, but didn't see the answer to this : ) I'd like to understand what is going on. Maybe there's something about the theme (I use [Paper](https://github.com/nanxiaobei/hugo-paper) as a base).

**Comment by jmooring** [comment]
@bolnel Please use the forum (https://discourse.gohugo.io/) for questions and troubleshooting. We prefer to use GitHub for verified bugs and vetted enhancements. Thanks.

**Comment by XhstormR** [comment]
If i use markdown syntax for images,  but we removed `canonifyURLs` support,  What's the best alternative solution for the baseURL with sub directory?

**Comment by jmooring** [comment]
You can enable Hugo's embedded link and image render hooks:

- <https://gohugo.io/render-hooks/links/#default>
- <https://gohugo.io/render-hooks/images/#default>


Note that these are enabled by default with multilingual single-host sites.

There is never a good reason to use `{{ site.BaseURL }}` in a template. At some point I hope we deprecate this method as well.

**Comment by ksgolding** [comment]
Would like to thank [XhstormR](https://github.com/XhstormR) for providing an actual solution to the problem! I have to admit, that how Hugo manages paths for images, well and apparently links is strangely fustrating and obtuse, especially for something that should really just work. I expect that the "static" folder is likely causing more problems then it solves.

Here is a subtly alternate to [XhstormR](https://github.com/XhstormR) solution, not using the apparently bad BaseURL

```
<img src='{{ urls.RelURL "" }}{{ .Destination | safeURL }}'
  {{- with .Text }} alt="{{ . }}"{{ end -}}
  {{- with .Title }} title="{{ . }}"{{ end -}}
>
{{- /* chomp trailing newline */ -}}
```

**Comment by jmooring** [comment]
@ksgolding 

> especially for something that should really just work

I've written about this many times, with a detailed description [here](https://www.veriphor.com/articles/link-and-image-render-hooks/). The short answer is to use link and image render hooks that resolve markdown destinations regardless of site configuration and front matter (e.g., `slug` and `url`) values.

The easiest way to handle this is to enable Hugo's embedded link and image render hooks:

- <https://gohugo.io/getting-started/configuration-markup/#renderhookslinkuseembedded>
- <https://gohugo.io/getting-started/configuration-markup/#renderhooksimageuseembedded>


You can examine the source code here:

- <https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/_default/_markup/render-link.html>
- <https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/_default/_markup/render-image.html>
